pkg/spanconfig/spanconfigsqlwatcher/buffer.go

```diff
@@ -119,27 +120,84 @@ func (b *buffer) flushEvents(
 // combined frontier timestamp is also returned.
 func (b *buffer) flush(
 	ctx context.Context,
-) (updates []spanconfig.DescriptorUpdate, _ hlc.Timestamp, _ error) {
+) (sqlUpdates []spanconfig.SQLUpdate, _ hlc.Timestamp, _ error) {
 	events, combinedFrontierTS := b.flushEvents(ctx)
+	// The events slice can contain a SQLUpdate that applies to a descriptor, or a
+	// SQLUpdate that applies to a protected timestamp record. We bubble up the
+	// protected timestamp SQLUpdates to the end of the slice, so that we can
+	// perform the subsequent deduplication of SQLUpdates. We do this instead of
+	// pre-processing events to separate the two kinds of updates, to save on the
+	// allocation of additional `events` slices.
 	sort.Slice(events, func(i, j int) bool {
 		ei, ej := events[i].(event), events[j].(event)
-		if ei.update.ID == ej.update.ID {
-			return ei.timestamp.Less(ej.timestamp)
+		if ei.update.IsDescriptorUpdate() && ej.update.IsDescriptorUpdate() {
+			descUpdatei := ei.update.GetDescriptorUpdate()
+			descUpdatej := ej.update.GetDescriptorUpdate()
+			if descUpdatei.ID == descUpdatej.ID {
+				return ei.timestamp.Less(ej.timestamp)
+			}
+			return descUpdatei.ID < descUpdatej.ID
 		}
-		return ei.update.ID < ej.update.ID
+
+		// If the LHS is a descriptor update, sort it before the RHS which is a
+		// protected timestamp update.
+		if ei.update.IsDescriptorUpdate() {
+			return true
+		}
+
+		// If the RHS is a descriptor update, sort it before the LHS which is a
+		// protected timestamp update.
+		if ej.update.IsDescriptorUpdate() {
+			return false
+		}
+
```

Comment:
Hmm we could sort cluster before tenant, but if two tenant targets had the same timestamp I'm unsure how we would provide a stable sort. Unless we compare the proto Strings or something.

Reply:
Ended up sorting tenant updates before cluster, but the sort still isnt stable if two tenants or two cluster targets have the same timestamp.

Reply:
Discussed a bit more offline, and also in relation to #74612 which this PR slightly overlaps with when sorting []rangefeedbuffer.Event. The proposal here was to represent an individual tenant protection as unit spanconfig.SQLUpdate, instead of representing it using an embedded slice, which would prevent sorting two elements of the type. Sorry for forgetting that we were using slices internally preventing the stable sort + necessitating the proto-string-for-map-key below.